syntax = "proto3";

message requestVote {
    int32 term = 1;
    int32 cid = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message requestVoteReply {
    int32 term = 1;
    bool voteGranted = 2;
    int32 id = 3;
    repeated log entries = 4;
    float leaderLeaseStartTime = 5;
}

message log {
    string operation = 1;
    string varName = 2;
    string varValue = 3;
    int32 term = 4;
}

message appendEntries {
    int32 term = 1;
    int32 leaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated log entries = 5;
    int32 leaderCommitIndex = 6;
}

message appendEntriesReply {
    int32 term = 1;
    bool success = 2;
}

message ServeClientArgs {
    string Request = 1;
}

message ServeClientReply {
  string Data = 1;
  int32 LeaderID = 2;
  bool Success = 3;
}

service Raft {
    rpc ElectionRequestVoteRPC(requestVote) returns (requestVoteReply) {}
    rpc HeartBeatAppendEntriesRPC(appendEntries) returns (appendEntriesReply) {}
    rpc ServeClient (ServeClientArgs) returns (ServeClientReply) {}
}
